accepted/2024/package-update-command.md

```diff
@@ -0,0 +1,206 @@
+This flag will try to update only packages that have direct or transitive vulnerabilities. The remediation is calculated with an implicit dotnet audit to then apply directly to a resulting package graph. It can add packages, remove packages, and update packages depending on the problem it's attempting to resolve. It does not take into consideration downgrading to a compatible version if a higher one has already been specified.
+The algorithm of the fixing process will have the following steps:
+
+1. Identify  vulnerabilities in the graph
+1. Distinguish which are top level, transitive level, etc.,
+1. Resolve any compatibility conflicts
+1. Apply the patch/update/fix
+1. Restore dependencies
+1. Verify that issues were resolved  . Provide the report to the user.
+1. If the issues were not resolved or were partially resolved, we will still commit the fixes and notify users about what was and was not fixed and what are the recommended manual steps with links to documentation.
+
+Example:
+
+```CLI
+C:\> dotnet package update --vulnerable
+
+Fetching package metadata from: 'https://api.nuget.org/v3/index.json'
+Loaded 23 security advisories from 'https://api.nuget.org/v3/index.json'
+Scanning ContosoUniversity.sln (36 NuGet dependencies)
+
+error: Vulnernable packages found!
+[net5.0]:
+Top-level Package     Requested Resolved Severity Advisory URL
+> UmbracoForms      8.4.1  8.4.1  Moderate https://github.com/advisories/GHSA-8m73-w2r2-6xxj
+
+Transitive Package  Resolved Severity Advisory URL
+> Microsoft.Data.OData 5.2.0  Moderate https://github.com/advisories/GHSA-mv2r-q4g5-j8q5
+
+Found 1 top-level Moderate severity vulnerability & 1 transtive Moderate severity vulnerability package(s) in 36 scanned packages.
+
+Fixing vulnernable packages in ContosoUniversity.sln
+ Upgrading UmbracoForms 8.4.1 to 8.7.1
+
+Fixed 2 packages in 36 scanned packages.
+```
+
+#### --mode
+
+Priority: P1 for `promote`; P2 for `closest` and switching the default to `closest`.
+
+| Mode | Explanation |
+|------|-------------|
+| `promote` | the algorithm will always promote transitive packages to top level. This is an easy to implement feature that we will release in MVP. |
+| `closest` | the algorithm will try to upgrade the direct package reference, and if that doesn't work, it will walk the graph one dependency level at a time trying to upgrade that, until it gets to the package with the known vulnerability. Once this mode is implemented, it should be set to be the default one. |
+
+#### --format json
+
+Priority: P2
+
+Provides a detailed report in a JSON format:
+
+```json
+  "added": [
+    {
+      "name": "Microsoft.AspNetCore.MVC",
+      "version": "2.2.0"
+    }
+  ],
+  "removed": [
+      {
+        "name": "anurse.testing.TestDeprecatedPackage",
+        "version": "1.0.0"
+      }
+  ],
+  "updated": [
+    {
+      "name": "UmbracoForms",
+      "version": "8.7.1",
+      "previousVersion": "8.4.1"
+    }
+  ],
+  "failures": [],
+  "warnings": []
+```
```

Comment:
Since this command is going to be used in solutions too, I think we need to be more detailed in this json format. For example, let's say that dotnet audit fix resolves to different package version for different projects, or for different target frameworks, I think we need to reflect that in the JSON output.

Another edge case I can think of is when customer has project references. Let's say that ProjectA has a reference to ProjectB that contains a vulnerable package, since those packages are considered transitive will the solution be updating the package in ProjectB and install it in ProjectA?

With those considerations I was thinking something like this:

```json

{
	"version": "string",
	"feasible": true,
	"metadata": 
	{
		"added": "number",
		"removed": "number",
		"updated": "number",
		"promotedToTopLevel": "number",
	},
	"vulnerabilities":
	{
		"projectA":
		{
			"net8.0":
			{
				"total": "number",
				"newtonsoft.json": 
				{
					"originalVersion": "12.0.1", // Could be ranges too
					"newVersion": "13.0.1", // Could be ranges too
					"promotedToTopLevel": true,
					"projectReference": true, // Could be a path to the project reference
					...
				},
				"castle.core":
				{
					...
				},
			},
			"net7.0":
			{
				...
			},	
		},
		"projectB":
		{
			...
		},
	},
}
```
@OliaG @zivkan What do you think about this JSON proposal?

Reply:
I have a more comments than I thought I would:

1. The proposed schema doesn't provide a place to output warning or error messages. If the command fails, the error should still be in a valid JSON, to make it easier to script. I believe that having an `errors` or `messages` array removes the need for a `feasible` bool, since any result without any errors is by definition feasible.
2. Some languages make it hard to use json properties that are not valid language variable names, so property keys like "net8.0" is problematic, as is using project names. Even when your language allows you to parse it, it can be harder to use since property/key names are dynamic. In some languages it's much easier when the schema is static. For example, consider trying to use json query to search the result, or use `Convert-FromJson` in PowerShell.
3. On the topic of project names, it's possible to have multiple projects in a solution have the same name, as long as they're in different subdirectories, so we have to use the full path to the project, not just the name, otherwise in edge cases they won't be unique.
4. In the case of a package being transitive through a project reference, I think the tool should be smart enough to not update the package. In that case, since the package isn't being updated, there's no need for a `"projectReference"` property.
5. `dotnet package update` will be used for all kinds of updates, not just vulnerable packages. I think Olia's `"added":`, `"updated":`, and `"removed":` properties make more sense than a `vulnerabilities` property key. Although in that property you have a collection of projects, so `projects` is a more appropriate name.
6. I haven't been looking at JSON other than NuGet's own for years, so I'm quite out of touch with JSON best practises. But I don't understand the purpose of putting the counts within a `metadata` object. Is there a benefit over putting the counts in the parent object?
7. I feel like the counts should be per-project, not at the root object level. Having it at the root object level raises questions, like if multiple projects have to promote the same package to top level, does that count as 1, or is the promote to top level count going to equal the number of projects?  However, I don't understand the point of the count properties, since you can look at look at the length of the array of package updates.
8. JSON numbers aren't quoted. Some languages/parsers might be loosey-goosey, but strict parsers will fail. I feel like NuGet had this problem when migrating one of our parsers from Newtonsoft.Json (using JToken) to System.Text.Json.
9. While nuget.org doesn't have a package with id `total`, it is a valid package ID, so it should not be used in a place where package IDs are also expressed. However, I already mentioned we shouldn't have a dynamic schema, so this concern should automatically be solved when addressing where to put package ids.
10. Similar to the "I'm not familiar with JSON best practises" comment I made earlier, I'm not sure if long property names `originalVersion`, `newVersion`, `promotedToTopLevel` are commonly used. `from` and `to` I think are easy enough to understand, as are `original` and `new`. Assuming we have some docs somewhere that explains what it means, I think `promoted` might be a good enough name. `promotedToTopLevel` might be guessable without docs. However, I just had the thought that since it's possible to tell the difference between an "addition" vs an "upgrade", there could be a property `wasTransitive`, and I think that will be more clear to people than `promoted`